Prometheus/ObservedExemplar.cs

```diff
@@ -0,0 +1,67 @@
+﻿namespace Prometheus;
+
+/// <summary>
+/// Internal representation of an Exemplar ready to be serialized.
+/// </summary>
+internal struct ObservedExemplar
+{
+    private const int MaxRunes = 128;
+    public static readonly ObservedExemplar Empty = new();
+
+    internal static INowProvder NowProvider = new RealNowProvider();
+
+    public Exemplar.LabelPair[]? Labels { get; private set; }
+    public double Val { get; private set; }
+    public double Timestamp { get; private set; }
+
+    public ObservedExemplar()
+    {
+        Labels = null;
+        Val = 0;
+        Timestamp = 0;
+    }
+
+   internal interface INowProvder
+    {
+        double Now();
+    }
+
+   private sealed class RealNowProvider : INowProvder
+   {
+       public double Now()
+       { 
+           return DateTimeOffset.Now.ToUnixTimeMilliseconds() / 1e3;
+       }
+   }
+    
+    public bool IsValid => Labels != null;
+
+    public void Update(Exemplar.LabelPair[] labels, double val)
+    {
+        var tally = 0;
+        for (var i = 0; i < labels.Length; i++)
+        {
+            tally += labels[i].RuneCount;
+            for (var j = 0; j < labels.Length; j++)
+            {
+                if (i == j) continue;
+                if (Equal(labels[i].KeyBytes, labels[j].KeyBytes))
+                    throw new ArgumentException("exemplar contains duplicate keys");
+            }
+        }
+
+        if (tally > MaxRunes)
+            throw new ArgumentException($"exemplar labels has {tally} runes, exceeding the limit of {MaxRunes}.");
```

Comment:
I am not very comfortable throwing "bad data" type of exception from this code, largely because it is often unexpected to the users of the library - they expect that telemetry APIs are not going to complain about data, it is seen as "harmless" code.

What is the impact if we skip the length check and emit over-length data?

Reply:
[The go implementation has a panic in the instrumentation codepath.](https://github.com/prometheus/client_golang/blob/07b1397deda7a534e8fb59598c6cd6226038ff09/prometheus/histogram.go#L1008-L1019) I feel we should treat it similarly here.

> What is the impact if we skip the length check and emit over-length data?

I Suspect warnings from the storage backends in the logs and the Exemplars just being dropped.